# Welcome to Lab: Experimental Design 👨‍🔬⚗️🧪

Before this lab section, you completed the "Hello Survey" based on the questions you generated with your groups.  This week, you will do an analysis of the results and explore techniques to do great experimental design in Python!

A few tips to remember:

- Refer to your **lecture notebook** and the **pandas cheat sheet** to help you out with the code!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

<hr style="color: #DD3403;">

In [ ]:
# Meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Daniel"
ca1_name = "Linh"
ca2_name = "Eeshan"


# Say hello to each other!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!
partner1_name = "David Mocianko"
partner1_netid = "dmoci2"
partner1_inYourGroupLastWeek = "yes"

partner2_name = "Shane Moncayl"
partner2_netid = "moncayo3"
partner1_inYourGroupLastWeek = "yes"

partner3_name = "Evan Walker"
partner3_netid = "evanaw3"
partner1_inYourGroupLastWeek = "yes"

<hr style="color: #DD3403;">

# Part 1: Hello Dataset Analysis

In lab last week, you created a series of questions that made up the "Hello Dataset".  Before lab, you completed this survey along with the entire course.  Today, you're going to do analysis on this dataset! :)


## Your Binary Question Analysis

You will be analyzing the **binary question** your group asked the class from `lab_pandas` -- record it below.  If multiple people from your group had different binary questions, decide which one to use and record that question below.

*(Are you taking RHET 105 this semester?)*

## Load the Hello Dataset

The "Hello Dataset" is available here:
```
https://waf.cs.illinois.edu/discovery/hello-fa22.csv
```

Use Python to load this dataset into a DataFrame called `df`:

In [10]:
df=pd.read_csv("https://waf.cs.illinois.edu/discovery/hello-fa22.csv")
df

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0          CA Ana  AL1                    2.0                    5.0   
1          CA Ram  AYR                    2.0                    5.0   
2          CA Xin  AYO                    1.0                    5.0   
3           Lekha  AYH                    2.0                    2.0   
4            Reni  AYD                    2.0                    5.0   
..            ...  ...                    ...                    ...   
519          Bill  AYC                    1.0                    5.0   
520       Renault  AYR                    2.0                    5.0   
521  Shivam Patel  AYC                    1.0                    3.0   
522         Colin  AYI                    1.0                    5.0   
523          Evan  AYG                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                            2-Jan         Sierra Mist   
1                                            2-Jan                 7UP   
2                                            2-Jan         Sierra Mist   
3                                              0.5                 7UP   
4                                            2-Jan                 7UP   
..                                             ...                 ...   
519                                          2-Jan         Sierra Mist   
520                                            0.5         Sierra Mist   
521                                            0.5                 7UP   
522                                          2-Jan         Sierra Mist   
523                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
0               iOS            Apple          Pork   On/Before June 30  ...   
1               iOS            Apple           NaN     On/After July 1  ...   
2               iOS            Apple          Beef   On/Before June 30  ...   
3               iOS           Orange          Pork     On/After July 1  ...   
4               iOS            Apple          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
519             iOS            Apple          Beef     On/After July 1  ...   
520         Android           Orange          Pork   On/Before June 30  ...   
521             iOS            Apple           NaN   On/Before June 30  ...   
522             iOS           Orange          Beef   On/Before June 30  ...   
523         Android           Orange          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
0                                  Cold brew   
1                               Black coffee   
2                                  starbucks   
3                        pumpkin spice latte   
4                                      Mocha   
..                                       ...   
519                                      NaN   
520                                     moca   
521                          Don’t have one    
522                                      tea   
523                                      Tea   

    What is your favorite mobile app?  What is your favorite subject?  \
0                            Snapchat                            Math   
1                      Google chrome                      Mathematics   
2                           Starbucks                      Psychology   
3                             twitter               english (writing)   
4                           Instagram                            Math   
..                                ...                             ...   
519                               NaN                             NaN   
520                           wechat                              cs    
521                           Youtube                  World History    
522

## Find The Variable for Your Question

The command `df.columns.tolist()` will **list all of the columns (variables)** in a DataFrame.  Use `df.columns.tolist()` and read the output until you find the column name for your binary question.

- You may see the message: `Output exceeds the __size limit__. Open the full output data __in a text editor__`
- If you do, click the **in a text editor** link to open the whole list in a new tab. :)

In [11]:
df.columns.tolist()

['Name',
 'Lab',
 'Number 1 or Number 2?',
 'Number 3 or Number 5?',
 'What is your preferred way of saying one half?',
 '7UP or Sierra Mist?',
 'Android or iOS?',
 'Apple or Orange?',
 'Beef or Pork?',
 'Your birthday is...',
 'Do you prefer to wear black or white?',
 'Boba or Coffee?',
 'Cake or Pie?',
 'Canes or Chick-fil-a?',
 'Do you pour milk or cereal first?',
 'Do you prefer chocolate or vanilla?',
 'Coffee or Tea?',
 'Coke or Pepsi?',
 'Cookies or Ice Cream?',
 'Do you prefer ice cream in a cup or in a cone?',
 'Do you prefer dad or mom?',
 'Marvel or DC?',
 'Are you living in a dorm or an apartment?',
 'Your last digit in your NetID...',
 'Were you born in an even or odd month of the year?',
 'Would you rather have the superpower of flight or invisibility?',
 'UIUC Football or Basketball Team?',
 'French Fries or Tater Tots?',
 'Heads or Tails?',
 'Hot Weather or Cold Weather?',
 'Hot Dogs or Hamburgers?',
 'In-State or Out-of-State/International',
 'Introvert or Extrovert?',

## Record the Variable for Your Question

In the cell below, record your question's column name in the variable `analysis_column`.  You'll need that exact column name to reference data from that cell.

In [14]:
analysis_column = "Are you taking RHET 105 this semester?"
analysis_column

'Are you taking RHET 105 this semester?'

### 🔬 Test Case Checkpoint 🔬

In [15]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "You do not have a DataFrame called `df`."
assert( analysis_column in df ), "Make certain your analysis_column is EXACTLY as it appears in the Hello Dataset."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Column Subset: View Only Your Question's Responses

To begin your question's analysis, you may want to **focus only on your question**.  In lecture, we have primarily focused on creating subsets of rows. However, you can also create a subset of columns!

To create a DataFrame with just a single column, you can use the syntax:
```
df[ ["column name"] ]
```
- Notice the **double set of square brackets**.  You need the set for `df`, and the second set for the column name.
- You can find this syntax under the "Subset Variables (Columns)" on your cheat sheet!

Us Python to display only the column you're analyzing.  This will make it easy to see what the two options were for your question:

In [16]:
df[["Are you taking RHET 105 this semester?"]]

Are you taking RHET 105 this semester?
0                                       No
1                                       No
2                                       No
3                                       No
4                                       No
..                                     ...
519                                     No
520                                     No
521                                     No
522                                     No
523                                     No

[524 rows x 1 columns]

## DataFrame: Group One

Create a DataFrame, `df1`, that contains everyone who picked the first option of your two answers.  *(Make sure to make this a subset of the original `df`, and not just your one column.)*

In [21]:
df1 = df[(df["Are you taking RHET 105 this semester?"]=="No")]
df1

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0          CA Ana  AL1                    2.0                    5.0   
1          CA Ram  AYR                    2.0                    5.0   
2          CA Xin  AYO                    1.0                    5.0   
3           Lekha  AYH                    2.0                    2.0   
4            Reni  AYD                    2.0                    5.0   
..            ...  ...                    ...                    ...   
519          Bill  AYC                    1.0                    5.0   
520       Renault  AYR                    2.0                    5.0   
521  Shivam Patel  AYC                    1.0                    3.0   
522         Colin  AYI                    1.0                    5.0   
523          Evan  AYG                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                            2-Jan         Sierra Mist   
1                                            2-Jan                 7UP   
2                                            2-Jan         Sierra Mist   
3                                              0.5                 7UP   
4                                            2-Jan                 7UP   
..                                             ...                 ...   
519                                          2-Jan         Sierra Mist   
520                                            0.5         Sierra Mist   
521                                            0.5                 7UP   
522                                          2-Jan         Sierra Mist   
523                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
0               iOS            Apple          Pork   On/Before June 30  ...   
1               iOS            Apple           NaN     On/After July 1  ...   
2               iOS            Apple          Beef   On/Before June 30  ...   
3               iOS           Orange          Pork     On/After July 1  ...   
4               iOS            Apple          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
519             iOS            Apple          Beef     On/After July 1  ...   
520         Android           Orange          Pork   On/Before June 30  ...   
521             iOS            Apple           NaN   On/Before June 30  ...   
522             iOS           Orange          Beef   On/Before June 30  ...   
523         Android           Orange          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
0                                  Cold brew   
1                               Black coffee   
2                                  starbucks   
3                        pumpkin spice latte   
4                                      Mocha   
..                                       ...   
519                                      NaN   
520                                     moca   
521                          Don’t have one    
522                                      tea   
523                                      Tea   

    What is your favorite mobile app?  What is your favorite subject?  \
0                            Snapchat                            Math   
1                      Google chrome                      Mathematics   
2                           Starbucks                      Psychology   
3                             twitter               english (writing)   
4                           Instagram                            Math   
..                                ...                             ...   
519                               NaN                             NaN   
520                           wechat                              cs    
521                           Youtube                  World History    
522

## DataFrame: Group Two

Create a new DataFrame, `df2`, that contains everyone who picked the other answer:

In [22]:
df2 = df[(df["Are you taking RHET 105 this semester?"]=="Yes")]
df2

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
32          Beomsu   AYD                    1.0                    5.0   
50           Yihao   AYS                    2.0                    5.0   
58        Matthew    AYP                    2.0                    3.0   
77            Reva   AL1                    2.0                    3.0   
79         Bernard   AYN                    1.0                    5.0   
99            Sean   AYS                    1.0                    3.0   
100          Kyle    AYJ                    1.0                    3.0   
114        Yuxiang   AYE                    1.0                    5.0   
115         Jeremy   AYO                    1.0                    3.0   
119       Sabarish   AYD                    2.0                    5.0   
138           Sean   AYJ                    2.0                    3.0   
140         Andrew   AYF                    1.0                    3.0   
154          Henri   LBD                    1.0                    5.0   
158           Sota   AYT                    1.0                    5.0   
173       Jonathan   AYR                    2.0                    3.0   
198            Tom   AYD                    2.0                    5.0   
208          Jimmy   AYA                    1.0                    3.0   
221       Scarlett   AYJ                    2.0                    5.0   
222          Grace   AYD                    1.0                    3.0   
225      zhenqiang   AYA                    1.0                    3.0   
233           Rama   AYL                    1.0                    5.0   
243           Evan   AYD                    1.0                    5.0   
248         Ishaan   AYR                    1.0                    3.0   
275      Seongwook   AYE                    2.0                    3.0   
283          Harry   AYE                    1.0                    3.0   
285  Daksheshkaran   AYP                    1.0                    5.0   
286          JACKY   AYA                    1.0                    3.0   
303       Pranavi    AYJ                    2.0                    5.0   
310      Changrong   AYL                    1.0                    5.0   
325        Stephen   AYN                    1.0                    3.0   
332          Lanbu  BYOD                    1.0                    3.0   
337            Ben   AYA                    1.0                    3.0   
344       Anushree   AYH                    2.0                    5.0   
346        Michael   AYI                    1.0                    3.0   
357         Rishab   AYG                    2.0                    3.0   
388          Tanya   AYC                    2.0                    3.0   
399           Feng  Stat                    1.0                    3.0   
421         Gianni   AYP                    1.0                    5.0   
436           Lizy   LBD                    2.0                    5.0   
450         Andrew   AYA                    1.0                    3.0   
452         Luyang   AYB                    1.0                    5.0   
467      Jeonghwan   AYL                    1.0                    3.0   
468           jake   LBD                    2.0                    5.0   
493          Yuang   AYF                    1.0                    3.0   
495          Aiden   AL1                    2.0                    3.0   
500            Ray   AYO                    2.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
32                                           2-Jan         Sierra Mist   
50                                           2-Jan         Sierra Mist   
58                                             0.5                 7UP   
77                                           2-Jan         Sierra Mist   
79                                           2-Jan                 7UP   
99                                           2-Jan                 7UP   
100              

## Analysis

Discuss the following questions with your group and then write down your response!

**Q1: What percentage of people chose each of your two binary options?**

*(476/522 people answered no. This translates to 91.2%. The remaining 46/522 people answered yes. This translates to 8.8%. The total amount of people who responded were 522. 524 people were asked. Two people didn't answer.)*

**Q2: Did this surprise you? Why or why not? Write at least one complete sentence.**

*(I was surprised by the answers that were provided. I felt that the results would be closer to a 50/50 split but the majority of people, almost all of them, weren't taking it this semester with only a few people taking it this semester. This was very surprising.)*

**Q3: Do you think your question was the closest to a perfect 50/50 split among the whole class?**

*(No)*

<hr style="color: #DD3403;">

## Part 2: Creating Treatment and Control Groups

For the rest of this lab, we will consider the "Hello Dataset" to be a pre-survey of randomly selected participants for an upcoming experiment.  In lecture, you learned that a good experiment will split the participants into two groups: the "treatment" group and the "control" group.

Recall that it is necessary to have both groups:

- The "control" group is a comparison group that will receive a placebo and will act as a control to ensure that the treatment has an effect.
- The "treatment" group will receive a treatment and we will look for differences between the results of this treatment group and the control group.

## Part 2.1: Select a Treatment Group

Select half (50%) of the participants to be in the treatment group and place them into a DataFrame called `treatment`.
- Remember, you can use your **pandas cheat sheet** to help you out!

In [23]:
treatment = df.sample(frac=0.5)
treatment

Name       Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
313   Jaypal       AYT                    1.0                    5.0   
187   Dhiraj       AYQ                    2.0                    5.0   
430   Karena       AL1                    2.0                    3.0   
509     Tess       AYC                    2.0                    5.0   
472  CA Linh  AYD, AYF                    1.0                    3.0   
..       ...       ...                    ...                    ...   
364   Junhao       AYN                    1.0                    3.0   
400  Jeffrey       AYS                    2.0                    5.0   
131  Sarthak       AYB                    2.0                    5.0   
81      Anne       AYP                    1.0                    5.0   
155    Grace       AYN                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
313                                            0.5                 7UP   
187                                          2-Jan                 7UP   
430                                          2-Jan         Sierra Mist   
509                                          2-Jan                 7UP   
472                                          2-Jan                 7UP   
..                                             ...                 ...   
364                                          2-Jan                 7UP   
400                                            0.5                 7UP   
131                                          2-Jan         Sierra Mist   
81                                             0.5         Sierra Mist   
155                                            0.5                 NaN   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
313             iOS            Apple          Pork   On/Before June 30  ...   
187             iOS            Apple          Beef   On/Before June 30  ...   
430             iOS            Apple          Beef   On/Before June 30  ...   
509             iOS            Apple          Beef   On/Before June 30  ...   
472             iOS           Orange          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
364             iOS            Apple          Beef   On/Before June 30  ...   
400             iOS           Orange          Beef     On/After July 1  ...   
131             iOS           Orange          Beef   On/Before June 30  ...   
81              iOS            Apple          Beef     On/After July 1  ...   
155         Android           Orange          Pork   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
313                                Starbucks   
187                                 Expresso   
430                              iced coffee   
509                                cold brew   
472                                      NaN   
..                                       ...   
364                                Starbucks   
400                           Regular coffee   
131                               London Fog   
81              Mango Dragon Fruit Refresher   
155                   honey lemon ginger tea   

    What is your favorite mobile app? What is your favorite subject?  \
313                           Netflix                       STAT 107   
187                          Snapchat                None of them :)   
430                         pinterest                            art   
509                           twitter                        History   
472                            TikTok                            NaN   
..                                ...                            ...   
364                            WeChat              Computer Science    
400                         Pinterest                           Math   
131                            Safari                        History   
81      

## Part 2.2: Select a Control Group

Using the **exact same technique and syntax** as you did above, also select half (50%) of the participants to be in the control group.  Place them into a DataFrame called `control`. For now, it's okay if there are people who are in both groups.

In [24]:
control = df.sample(frac=0.5)
control

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
491       Minghan  AL1                    2.0                    3.0   
521  Shivam Patel  AYC                    1.0                    3.0   
346       Michael  AYI                    1.0                    3.0   
481          Abhi  AYA                    1.0                    5.0   
486      julissa   lbd                    1.0                    5.0   
..            ...  ...                    ...                    ...   
446          John  AYA                    2.0                    3.0   
268           Max  AYL                    1.0                    3.0   
321            Ji  AYQ                    1.0                    3.0   
391         Brian  AYK                    2.0                    3.0   
53       Xiangyi   AYP                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
491                                            0.5         Sierra Mist   
521                                            0.5                 7UP   
346                                            0.5                 7UP   
481                                          2-Jan                 7UP   
486                                          2-Jan                 7UP   
..                                             ...                 ...   
446                                          2-Jan                 NaN   
268                                          2-Jan                 7UP   
321                                            0.5                 7UP   
391                                          2-Jan                 7UP   
53                                           2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
491             iOS            Apple          Beef   On/Before June 30  ...   
521             iOS            Apple           NaN   On/Before June 30  ...   
346             iOS            Apple          Beef   On/Before June 30  ...   
481             iOS           Orange           NaN     On/After July 1  ...   
486             iOS           Orange          Beef   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
446             iOS           Orange          Pork   On/Before June 30  ...   
268             iOS           Orange          Beef   On/Before June 30  ...   
321             iOS           Orange          Beef     On/After July 1  ...   
391             iOS            Apple          Beef   On/Before June 30  ...   
53              iOS           Orange          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
491                                starbucks   
521                          Don’t have one    
346                                    Water   
481                                   coffee   
486                            iced milk tea   
..                                       ...   
446                                cold brew   
268                               frappecino   
321                                Starbucks   
391                            Vanilla Latte   
53                                     latte   

    What is your favorite mobile app? What is your favorite subject?  \
491                            wechat                             IS   
521                           Youtube                 World History    
346                         Instagram                           Math   
481                               NaN                            NaN   
486                         pinterest                        English   
..                                ...                            ...   
446                            tiktok                       history    
268                           stock x                           math   
321                            Reddit                           Math   
391          

### 🔬 Test Case Checkpoint 🔬

In [25]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "You do not have a DataFrame called `df`."
assert( 'treatment' in vars() ), "You do not have a DataFrame called `treatment`."
assert( 'control' in vars() ), "You do not have a DataFrame called `control`."
assert( len(treatment) == len(control) ), "The `treatment` and `control` groups must be the same size."
assert( abs((len(df) / 2) - len(treatment)) <= 1 ), "The `treatment` group size must be 50% of the `df`"
assert( abs((len(df) / 2) - len(control)) <= 1 ), "The `control` group size must be 50% of the `df`"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Analysis: Overlapping Selection?

The following code will detect rows that are **both in your treatment and control** groups.

Run the following cell to see everyone who is in both groups:

In [26]:
df[ df.index.isin(treatment.index) & df.index.isin(control.index) ]

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
11         Ashley  AYN                    2.0                    5.0   
13        Anvitha  AYP                    1.0                    5.0   
14      Tianxiang  AYR                    1.0                    5.0   
15         Steven  AYR                    1.0                    3.0   
24          Brian  AYF                    1.0                    3.0   
..            ...  ...                    ...                    ...   
513           Jen  AYC                    2.0                    5.0   
514          Bill  AYA                    1.0                    3.0   
515        Faayez  AYT                    2.0                    5.0   
516        Edward  AYQ                    1.0                    5.0   
521  Shivam Patel  AYC                    1.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
11                                             0.5         Sierra Mist   
13                                             0.5                 7UP   
14                                           2-Jan                 7UP   
15                                           2-Jan         Sierra Mist   
24                                             0.5                 7UP   
..                                             ...                 ...   
513                                          2-Jan                 7UP   
514                                          2-Jan                 7UP   
515                                          2-Jan                 7UP   
516                                            0.5         Sierra Mist   
521                                            0.5                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
11              iOS            Apple          Beef   On/Before June 30  ...   
13              iOS            Apple          Beef     On/After July 1  ...   
14          Android           Orange          Pork     On/After July 1  ...   
15              iOS            Apple          Beef   On/Before June 30  ...   
24              iOS            Apple          Beef   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
513         Android            Apple          Pork     On/After July 1  ...   
514             iOS           Orange          Beef   On/Before June 30  ...   
515         Android            Apple          Beef   On/Before June 30  ...   
516             iOS           Orange          Beef     On/After July 1  ...   
521             iOS            Apple           NaN   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
11                         Caramel Macchiato   
13                                any coffee   
14                                 Starbucks   
15                                Frappucino   
24                                Cappuccino   
..                                       ...   
513                                     None   
514                                 Starbuck   
515                               Chai latte   
516                                    Philz   
521                          Don’t have one    

    What is your favorite mobile app? What is your favorite subject?  \
11                             Sudoku                           Math   
13                         instagram                            math   
14                             Wechat                             CS   
15                       Chipotle App                             CS   
24                            Youtube                           Math   
..                                ...                            ...   
513                           Twitter                           Art    
514                           Discord                           STAT   
515                           Spotify                      Economics   
516          

**Q: Why would your code result in some people being included in both groups? Write at least one complete sentence explaining why you think this might occur.**

*(✏️ The code will result in some people being in both groups because we have two individual statements that randomly select people from the original set of data. The people who are in group 1 aren't removed out of the dataset when we randomly select for group 2. ✏️)*

## Part 2.3:

Read the following guide on **using `.index.isin` to select all rows from a DataFrame not in another DataFrame**: https://discovery.cs.illinois.edu/guides/DataFrame-Row-Selection/dataframe-isin-selection/

Finally, create a new `control` group that contains all of the participants from `df` that are not in `treatment`:

In [36]:
control = df[~df.index.isin(treatment.index)]
control

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0    CA Ana  AL1                    2.0                    5.0   
1    CA Ram  AYR                    2.0                    5.0   
2    CA Xin  AYO                    1.0                    5.0   
3     Lekha  AYH                    2.0                    2.0   
4      Reni  AYD                    2.0                    5.0   
..      ...  ...                    ...                    ...   
507   Laiba  AYO                    1.0                    5.0   
508   Julio  AYC                    2.0                    5.0   
519    Bill  AYC                    1.0                    5.0   
522   Colin  AYI                    1.0                    5.0   
523    Evan  AYG                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                            2-Jan         Sierra Mist   
1                                            2-Jan                 7UP   
2                                            2-Jan         Sierra Mist   
3                                              0.5                 7UP   
4                                            2-Jan                 7UP   
..                                             ...                 ...   
507                                            0.5                 7UP   
508                                          2-Jan         Sierra Mist   
519                                          2-Jan         Sierra Mist   
522                                          2-Jan         Sierra Mist   
523                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
0               iOS            Apple          Pork   On/Before June 30  ...   
1               iOS            Apple           NaN     On/After July 1  ...   
2               iOS            Apple          Beef   On/Before June 30  ...   
3               iOS           Orange          Pork     On/After July 1  ...   
4               iOS            Apple          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
507             iOS            Apple          Beef     On/After July 1  ...   
508             iOS           Orange          Pork     On/After July 1  ...   
519             iOS            Apple          Beef     On/After July 1  ...   
522             iOS           Orange          Beef   On/Before June 30  ...   
523         Android           Orange          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
0                                  Cold brew   
1                               Black coffee   
2                                  starbucks   
3                        pumpkin spice latte   
4                                      Mocha   
..                                       ...   
507                                Starbucks   
508                             Black coffee   
519                                      NaN   
522                                      tea   
523                                      Tea   

    What is your favorite mobile app?  What is your favorite subject?  \
0                            Snapchat                            Math   
1                      Google chrome                      Mathematics   
2                           Starbucks                      Psychology   
3                             twitter               english (writing)   
4                           Instagram                            Math   
..                                ...                             ...   
507                          Snapchat                              cs   
508                         Instagram                         English   
519                               NaN                             NaN   
522                           YouTube                            Math

### 🔬 Test Case Checkpoint 🔬

In [37]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "You do not have a DataFrame called `df`."
assert( 'treatment' in vars() ), "You do not have a DataFrame called `treatment`."
assert( 'control' in vars() ), "You do not have a DataFrame called `control`."
assert( abs((len(df) / 2) - len(treatment)) <= 1 ), "The `treatment` group size must be 50% of the `df`"
assert( abs((len(df) / 2) - len(control)) <= 1 ), "The `control` group size must be 50% of the `df`"
assert( len(df[ df.index.isin(treatment.index) & df.index.isin(control.index) ]) == 0), "The `treatment` and `control` groups have some overlap!  There must be no overlap."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 2.4: Finding Yourself!

Use the following cells to write Python code to find yourself in either the treatment group or control group.
- **Hint**: How do you select a subset of rows?  What conditional will identify you??
- If you did not fill out the "Hello Survey" then you are not in this dataset. :(
- If you are not in this survey, find either "Prof. Wade" or "Prof. Karle" instead.  *(Those are the names we used in the `Name` column.)*

In [43]:
me = treatment[treatment.Name == "Sabarish"]
me

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
119  Sabarish  AYD                    2.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
119                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
119             iOS            Apple          Pork   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
119         Java Chip Frappacino - Starbucks   

    What is your favorite mobile app? What is your favorite subject?  \
119                         Instagram                           Math   

    Who is your favorite singer? What is your favorite season?  \
119                   Bruno Mars                          Fall   

    What is your Zodiac sign? What is your favorite video game?  \
119                     Virgo                             Forza   

    What is your favorite non-video game? What is your favorite food?  \
119                              Monopoly               Chicken Wings   

    What's your favorite movie?  
119         Wolf of Wall Street  

[1 rows x 120 columns]

## Analysis: Finding Yourself!

**Q: Which group where you in -- treatment or control?**


*(✏️ Treatment ✏️)*

<hr style="color: #DD3403;">

# Part 3: Using Blocking to Create Treatment and Control Groups that are Balanced

In your pre-experiment research, you have reason to believe that how the participants answer **your group's question** might influence the response of this experiment!  Oh no! Even though this is a large sample of participants, we want to block to be absolutely sure that the treatment and control groups are balanced with regard to your group's question!

To control for this, we will using **blocking**!  Remember from lecture that blocking has us creating subgroups for each response, selecting equal-sized treatment and control groups from each block, and then combining all the treatment groups together and all the control groups together to get the final groups. This ensures that your groups are balanced with regard to the variable you blocked on!

You have already done some of this work:
- Earlier in this notebook, you created `df1` that includes everyone who answered your group's question one way.
- Earlier in this notebook, you created `df2` that includes everyone who answered your group's question the other way.

## Part 3.1:

Create a `df1_treatment` group that contains a random selection of 50% of the rows from `df1`.

In [46]:
df1_treatment = df1.sample(frac=0.5)
df1_treatment

Name       Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
152       Will       AYS                    2.0                    5.0   
456         Mo       AYA                    2.0                    5.0   
196       Yihe       AYD                    2.0                    3.0   
252     Oliver       AYS                    1.0                    5.0   
151  TA Daniel  AYB, AYD                    1.0                    5.0   
..         ...       ...                    ...                    ...   
269      Kevin       LBD                    1.0                    3.0   
74       Alvin       AYT                    2.0                    5.0   
335         SG       AYF                    2.0                    5.0   
239       Jian       AYS                    1.0                    3.0   
163     Corina       AYT                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
152                                          2-Jan                 7UP   
456                                          2-Jan                 7UP   
196                                          2-Jan         Sierra Mist   
252                                          2-Jan                 7UP   
151                                          2-Jan         Sierra Mist   
..                                             ...                 ...   
269                                          2-Jan                 7UP   
74                                           2-Jan         Sierra Mist   
335                                            0.5                 7UP   
239                                            0.5         Sierra Mist   
163                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
152             iOS            Apple          Beef   On/Before June 30  ...   
456             iOS           Orange          Beef   On/Before June 30  ...   
196             iOS           Orange          Beef   On/Before June 30  ...   
252             iOS           Orange          Beef     On/After July 1  ...   
151         Android            Apple          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
269             iOS            Apple          Beef   On/Before June 30  ...   
74          Android            Apple          Beef     On/After July 1  ...   
335             iOS           Orange          Beef     On/After July 1  ...   
239             iOS           Orange          Pork     On/After July 1  ...   
163             iOS           Orange          Beef   On/Before June 30  ...   

              What is your favorite coffee shop drink?  \
152                                           Chai Tea   
456                                         Chai latte   
196                                          Americano   
252                                         Chai Latte   
151                                               Boba   
..                                                 ...   
269                                             Dunkin   
74                              Green Tea Frappuccino    
335  Iced Decaf Coffee with 2 Sweet n’ Low and Coco...   
239                                        Dulce Latte   
163                                          Cafe Kopi   

    What is your favorite mobile app? What is your favorite subject?  \
152                          Transit                            Stat   
456                            TikTok                       Business   
196                           Spotify                           STAT   
252                           Discord               Computer Science   
151                           Discord            Atmospheric Science   
..                                ...                            ...   
269                          Snapchat               Computer Science   
74        

## Part 3.2:

Create a `df1_control` group that contains the other half of the rows from `df1`. *(You can refer back to Part 2 to refresh on how to do this, make sure you use `df1` and `df1_control` as your DataFrames.)*

In [47]:
df1_control = df1[~df1.index.isin(df1_treatment.index)]
df1_control

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
3           Lekha   AYH                    2.0                    2.0   
6        Esbeida    AYA                    2.0                    5.0   
7         Jocelyn   AYS                    1.0                    5.0   
8         Suvinay   AYP                    1.0                    5.0   
10        Maxwell   AYJ                    1.0                    5.0   
..            ...   ...                    ...                    ...   
516        Edward   AYQ                    1.0                    5.0   
517        Lauren   AYP                    1.0                    3.0   
519          Bill   AYC                    1.0                    5.0   
521  Shivam Patel   AYC                    1.0                    3.0   
523          Evan   AYG                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
3                                              0.5                 7UP   
6                                            2-Jan                 7UP   
7                                            2-Jan         Sierra Mist   
8                                            2-Jan                 7UP   
10                                           2-Jan                 7UP   
..                                             ...                 ...   
516                                            0.5         Sierra Mist   
517                                          2-Jan                 7UP   
519                                          2-Jan         Sierra Mist   
521                                            0.5                 7UP   
523                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
3               iOS           Orange          Pork     On/After July 1  ...   
6               iOS           Orange          Beef     On/After July 1  ...   
7               iOS           Orange          Beef     On/After July 1  ...   
8           Android           Orange           NaN     On/After July 1  ...   
10              iOS           Orange          Beef     On/After July 1  ...   
..              ...              ...           ...                 ...  ...   
516             iOS           Orange          Beef     On/After July 1  ...   
517             iOS            Apple          Beef   On/Before June 30  ...   
519             iOS            Apple          Beef     On/After July 1  ...   
521             iOS            Apple           NaN   On/Before June 30  ...   
523         Android           Orange          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
3                        pumpkin spice latte   
6                       Iced vanilla coffee    
7                Caramel crunch Frappuccino    
8                                Frappuccino   
10                  Peach green tea lemonade   
..                                       ...   
516                                    Philz   
517                                      tea   
519                                      NaN   
521                          Don’t have one    
523                                      Tea   

    What is your favorite mobile app?  What is your favorite subject?  \
3                             twitter               english (writing)   
6                           Snapchat                             Math   
7                             Spotify                         History   
8                             Netflix                       Astronomy   
10                            Spotify                            Math   
..                                ...                             ...   
516                       Xiaohongshu                         Finance   
517                         instagram                         english   
519                               NaN                           

### 🔬 Test Case Checkpoint 🔬

In [48]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df1_treatment' in vars() ), "You do not have a DataFrame called `df1_treatment`."
assert( 'df1_control' in vars() ), "You do not have a DataFrame called `df1_control`."
assert( len(df[ df.index.isin(df1_treatment.index) & df.index.isin(df1_control.index) ]) == 0), "The `df1_treatment` and `df1_control` groups have some overlap!  There must be no overlap."
assert( len(df[ df.index.isin(df1_treatment.index) & df.index.isin(df2.index) ]) == 0), "The `df1_treatment` and `df2` groups have some overlap!  There must be no overlap."
assert( len(df[ df.index.isin(df1_control.index) & df.index.isin(df2.index) ]) == 0), "The `df1_control` and `df2` groups have some overlap!  There must be no overlap."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 3.3:

Create a `df2_treatment` group that contains a random selection of 50% of the rows from `df2`.

In [49]:
df2_treatment = df2.sample(frac=0.5)
df2_treatment

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
495     Aiden   AL1                    2.0                    3.0   
114   Yuxiang   AYE                    1.0                    5.0   
357    Rishab   AYG                    2.0                    3.0   
221  Scarlett   AYJ                    2.0                    5.0   
346   Michael   AYI                    1.0                    3.0   
173  Jonathan   AYR                    2.0                    3.0   
450    Andrew   AYA                    1.0                    3.0   
233      Rama   AYL                    1.0                    5.0   
332     Lanbu  BYOD                    1.0                    3.0   
119  Sabarish   AYD                    2.0                    5.0   
243      Evan   AYD                    1.0                    5.0   
344  Anushree   AYH                    2.0                    5.0   
283     Harry   AYE                    1.0                    3.0   
32     Beomsu   AYD                    1.0                    5.0   
99       Sean   AYS                    1.0                    3.0   
77       Reva   AL1                    2.0                    3.0   
325   Stephen   AYN                    1.0                    3.0   
140    Andrew   AYF                    1.0                    3.0   
436      Lizy   LBD                    2.0                    5.0   
222     Grace   AYD                    1.0                    3.0   
50      Yihao   AYS                    2.0                    5.0   
208     Jimmy   AYA                    1.0                    3.0   
58   Matthew    AYP                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
495                                          2-Jan                 7UP   
114                                          2-Jan                 7UP   
357                                          2-Jan                 7UP   
221                                            NaN                 NaN   
346                                            0.5                 7UP   
173                                            0.5                 7UP   
450                                          2-Jan                 7UP   
233                                          2-Jan                 7UP   
332                                          2-Jan                 7UP   
119                                          2-Jan                 7UP   
243                                          2-Jan         Sierra Mist   
344                                            0.5         Sierra Mist   
283                                            0.5                 7UP   
32                                           2-Jan         Sierra Mist   
99                                           2-Jan                 7UP   
77                                           2-Jan         Sierra Mist   
325                                          2-Jan         Sierra Mist   
140                                            0.5         Sierra Mist   
436                                          2-Jan                 7UP   
222                                          2-Jan                 7UP   
50                                           2-Jan         Sierra Mist   
208                                          2-Jan                 7UP   
58                                             0.5                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
495         Android            Apple          Beef   On/Before June 30  ...   
114         Android           Orange          Beef   On/Before June 30  ...   
357             iOS           Orange          Beef     On/After July 1  ...   
221             iOS           Orange          Beef     On/After July 1  ...   
346             iOS            Apple          Beef   On/Before June 30  ...   
173         Android            Apple          Beef     On/After July 1  ...   
450             iOS   

## Part 3.4:

Create a `df2_control` group that contains the other half of the rows from `df2`.

In [50]:
df2_control = df2[~df2.index.isin(df2_treatment.index)]
df2_control

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
79         Bernard   AYN                    1.0                    5.0   
100          Kyle    AYJ                    1.0                    3.0   
115         Jeremy   AYO                    1.0                    3.0   
138           Sean   AYJ                    2.0                    3.0   
154          Henri   LBD                    1.0                    5.0   
158           Sota   AYT                    1.0                    5.0   
198            Tom   AYD                    2.0                    5.0   
225      zhenqiang   AYA                    1.0                    3.0   
248         Ishaan   AYR                    1.0                    3.0   
275      Seongwook   AYE                    2.0                    3.0   
285  Daksheshkaran   AYP                    1.0                    5.0   
286          JACKY   AYA                    1.0                    3.0   
303       Pranavi    AYJ                    2.0                    5.0   
310      Changrong   AYL                    1.0                    5.0   
337            Ben   AYA                    1.0                    3.0   
388          Tanya   AYC                    2.0                    3.0   
399           Feng  Stat                    1.0                    3.0   
421         Gianni   AYP                    1.0                    5.0   
452         Luyang   AYB                    1.0                    5.0   
467      Jeonghwan   AYL                    1.0                    3.0   
468           jake   LBD                    2.0                    5.0   
493          Yuang   AYF                    1.0                    3.0   
500            Ray   AYO                    2.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
79                                           2-Jan                 7UP   
100                                            0.5         Sierra Mist   
115                                          2-Jan                 7UP   
138                                          2-Jan         Sierra Mist   
154                                          2-Jan         Sierra Mist   
158                                          2-Jan                 7UP   
198                                            0.5                 7UP   
225                                          2-Jan         Sierra Mist   
248                                          2-Jan                 7UP   
275                                          2-Jan                 7UP   
285                                            0.5                 7UP   
286                                          2-Jan                 7UP   
303                                          2-Jan                 7UP   
310                                          2-Jan         Sierra Mist   
337                                          2-Jan         Sierra Mist   
388                                          2-Jan                 7UP   
399                                          2-Jan                 7UP   
421                                          2-Jan         Sierra Mist   
452                                            0.5         Sierra Mist   
467                                          2-Jan                 7UP   
468                                          2-Jan                 7UP   
493                                          2-Jan         Sierra Mist   
500                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
79              iOS           Orange          Beef   On/Before June 30  ...   
100             iOS            Apple          Beef   On/Before June 30  ...   
115             iOS            Apple          Beef   On/Before June 30  ...   
138             iOS            Apple          Beef     On/After July 1  ...   
154             iOS            Apple          Beef   On/Before Ju

### 🔬 Test Case Checkpoint 🔬

In [51]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df2_treatment' in vars() ), "You do not have a DataFrame called `df2_treatment`."
assert( 'df2_control' in vars() ), "You do not have a DataFrame called `df2_control`."
assert( len(df[ df.index.isin(df2_treatment.index) & df.index.isin(df2_control.index) ]) == 0), "The `df2_treatment` and `df2_control` groups have some overlap!  There must be no overlap."
assert( len(df[ df.index.isin(df2_treatment.index) & df.index.isin(df1.index) ]) == 0), "The `df2_treatment` and `df1` groups have some overlap!  There must be no overlap."
assert( len(df[ df.index.isin(df2_control.index) & df.index.isin(df1.index) ]) == 0), "The `df2_control` and `df1` groups have some overlap!  There must be no overlap."


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 3.5:

Finally, read the guide: **Combining DataFrames by Concatenation** (https://discovery.cs.illinois.edu/guides/Combining-DataFrames/Combining-DataFrames-by-Concatenation/) to learn how to combine `df1_treatment` and `df2_treatment` together into one group called `treatment`:

In [58]:
treatment = pd.concat([df1_treatment,df2_treatment])
treatment

Name       Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
152       Will       AYS                    2.0                    5.0   
456         Mo       AYA                    2.0                    5.0   
196       Yihe       AYD                    2.0                    3.0   
252     Oliver       AYS                    1.0                    5.0   
151  TA Daniel  AYB, AYD                    1.0                    5.0   
..         ...       ...                    ...                    ...   
436       Lizy       LBD                    2.0                    5.0   
222      Grace       AYD                    1.0                    3.0   
50       Yihao       AYS                    2.0                    5.0   
208      Jimmy       AYA                    1.0                    3.0   
58    Matthew        AYP                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
152                                          2-Jan                 7UP   
456                                          2-Jan                 7UP   
196                                          2-Jan         Sierra Mist   
252                                          2-Jan                 7UP   
151                                          2-Jan         Sierra Mist   
..                                             ...                 ...   
436                                          2-Jan                 7UP   
222                                          2-Jan                 7UP   
50                                           2-Jan         Sierra Mist   
208                                          2-Jan                 7UP   
58                                             0.5                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
152             iOS            Apple          Beef   On/Before June 30  ...   
456             iOS           Orange          Beef   On/Before June 30  ...   
196             iOS           Orange          Beef   On/Before June 30  ...   
252             iOS           Orange          Beef     On/After July 1  ...   
151         Android            Apple          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
436             iOS            Apple          Beef   On/Before June 30  ...   
222             iOS            Apple          Beef     On/After July 1  ...   
50              iOS           Orange          Beef     On/After July 1  ...   
208             iOS            Apple          Beef   On/Before June 30  ...   
58              iOS           Orange          Beef   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
152                                 Chai Tea   
456                               Chai latte   
196                                Americano   
252                               Chai Latte   
151                                     Boba   
..                                       ...   
436                               Iced Latte   
222                                  dunkin    
50                             Ice Americano   
208                                    Latte   
58                         Caramel Macchiato   

    What is your favorite mobile app? What is your favorite subject?  \
152                          Transit                            Stat   
456                            TikTok                       Business   
196                           Spotify                           STAT   
252                           Discord               Computer Science   
151                           Discord            Atmospheric Science   
..                                ...                            ...   
436                          Tik Tok                          STATS    
222                           youtube                           Math   
50                    Little red book                     

## Part 3.6:

Using the same technique, combine `df1_control` and `df2_control` together into one group called `control`:

In [57]:
control = pd.concat([df1_control,df2_control])
control

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
3        Lekha   AYH                    2.0                    2.0   
6     Esbeida    AYA                    2.0                    5.0   
7      Jocelyn   AYS                    1.0                    5.0   
8      Suvinay   AYP                    1.0                    5.0   
10     Maxwell   AYJ                    1.0                    5.0   
..         ...   ...                    ...                    ...   
452     Luyang   AYB                    1.0                    5.0   
467  Jeonghwan   AYL                    1.0                    3.0   
468       jake   LBD                    2.0                    5.0   
493      Yuang   AYF                    1.0                    3.0   
500        Ray   AYO                    2.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
3                                              0.5                 7UP   
6                                            2-Jan                 7UP   
7                                            2-Jan         Sierra Mist   
8                                            2-Jan                 7UP   
10                                           2-Jan                 7UP   
..                                             ...                 ...   
452                                            0.5         Sierra Mist   
467                                          2-Jan                 7UP   
468                                          2-Jan                 7UP   
493                                          2-Jan         Sierra Mist   
500                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
3               iOS           Orange          Pork     On/After July 1  ...   
6               iOS           Orange          Beef     On/After July 1  ...   
7               iOS           Orange          Beef     On/After July 1  ...   
8           Android           Orange           NaN     On/After July 1  ...   
10              iOS           Orange          Beef     On/After July 1  ...   
..              ...              ...           ...                 ...  ...   
452             iOS            Apple          Beef   On/Before June 30  ...   
467             iOS           Orange          Beef     On/After July 1  ...   
468             iOS            Apple          Beef     On/After July 1  ...   
493             iOS            Apple          Beef     On/After July 1  ...   
500         Android            Apple          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
3                        pumpkin spice latte   
6                       Iced vanilla coffee    
7                Caramel crunch Frappuccino    
8                                Frappuccino   
10                  Peach green tea lemonade   
..                                       ...   
452                                 Starbuck   
467                                Starbucks   
468                             black coffee   
493                                Starbucks   
500                                    water   

    What is your favorite mobile app? What is your favorite subject?  \
3                             twitter              english (writing)   
6                           Snapchat                            Math   
7                             Spotify                        History   
8                             Netflix                      Astronomy   
10                            Spotify                           Math   
..                                ...                            ...   
452                            TikTok                        history   
467                           YouTube                             IS   
468                           tic tok                           math   
493                            Wec

### 🔬 Test Case Checkpoint 🔬

In [56]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( len(df[ df.index.isin(treatment.index) & df.index.isin(control.index) ]) == 0), "The `treatment` and `control` groups have some overlap!  There must be no overlap."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Analysis: Blocking

**Q: In your own words, explain what blocking is and why it is useful to someone who has never taken a statistics course before! You can use what you did above as an example in your explanation if you'd like. Make sure to include what blocking is and why it is useful. Write at least three complete sentences to answer this question.**

*(✏️ Blocking is a method used to eliminate bias from small groups of data. The process involves categorizing the data into like-groups and then deciding who gets the treatment and control within each group. For instance, We can take a sample classroom and block the classmates based on age or gender to make the groups more alike and homogenous, eliminating bias and differences. ✏️)*

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)